### ISSUE TITLE
Using Windows Terminal with a git-for-windows tab displays odd characters

### ISSUE THREAD

**Issue by briannipper** [git-for-windows/git#2483]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
git version 2.24.1.windows.2
cpu: x86_64
built from commit: 992f0773022527b1b0cb1e0c13aec97dd5248053
sizeof-long: 4
sizeof-size_t: 8 
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Microsoft Windows [Version 10.0.18363.592]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
from C:\Program Files\Git\etc\install-options.txt
Editor Option: VIM
Custom Editor Path: 
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
Enable Builtin Interactive Add: Disabled
```

 * Any other interesting things about your environment that might be related
   to the issue you're seeing?
    * I've installed arcanist to interact with phabricator and when using the "standard" bash window I have no display issues; however, when I use the Bash window as a tab in Windows Terminal I have an unexpected issue. Here is a link to arcanist in case it's helpful:
https://secure.phabricator.com/book/phabricator/article/arcanist/

### Details

 * Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other
   * Windows Terminal

 * What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.
   * The following commands result in showing the issue which are all Arcanist commands
     * arc diff
     * arc commit
     * arc feature

 * What did you expect to occur after running these commands?
    * The command would run without showing any unexpected characters

 * What actually happened instead?
    * The commands run; however, there are clearly unexpected characters.

 * If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?
    * No specific URL for a repo; however, I do have a link to a bug related to Windows Terminal where they explain why this is really an issue with msys/mingw/cygwin runtime
https://github.com/microsoft/terminal/issues/4076#issuecomment-571308395

**Comment by dscho** [comment]
> The following commands result in showing the issue which are all Arcanist commands
> 
>     * arc diff
> 
>     * arc commit
> 
>     * arc feature

1. This is not a minimal example
2. This example is about Arcanist, not about Git for Windows. If you can replicate with bare GIt for Windows, please do so and provide the details here.

**Comment by briannipper** [comment]
@dscho
As I understand the problem based on the comments from the bug I opened (and was later closed) under the Microsoft Terminal the issue lies in...
msys/mingw/cygwin runtime

As such I was offering up the example of the commands I run in arcanist which are attempting to use aspects of color and text formating that result in odd characters.

As i listed in the other bug report with screen shots (and replicated here) I'm looking for help in addressing the way the odd characters appear in the msys/mingw/cygwin when it's used within Windows Terminal... perhaps this isn't an issue with git-for-windows then?

Image of mingw in windows terminal with odd characters:
https://user-images.githubusercontent.com/361720/71563235-1ff82200-2a5a-11ea-8f66-8aa5e3bb0b82.jpg

Image of PS Core in windows terminal running same commands without odd characters:
https://user-images.githubusercontent.com/361720/71563236-1ff82200-2a5a-11ea-8d2c-b467bd7c3f27.jpg

If this repo is ALSO not the right place to list this bug advise and I'll keep looking for the right place.

Thank you for your time.

**Comment by dscho** [comment]
I'll need a minimal reproducer to assess (and possibly fix) the issue. Not screenshots. I will not install arcanist and hunt for that reproducer myself.

**Comment by rimrul** [comment]
Ok, this issue seems to consist of multiple layers.

* you're probably calling the wrong bash
* you're probably using the wrong php
* the way arcanist determines wether they can use ansi colors is broken

Let me explain. You're probably using the recommended profile for Git Bash from the microsoft/terminal repo. That calls `bin/bash.exe` instead of `git-bash.exe` among other differences, this sets `TERM` to `cygwin` instead of `xterm`. That's the reason you can't reproduce this in standalone Git Bash.

You're probably also using a PHP for windows, not msys2 or cygwin. That means `DIRECTORY_SEPARATOR` is probably set to `\` and `posix_isatty()` doesn't exist. It also doesn't handle ANSI color sequences.

arcanist disables ANSI color sequences if `DIRECTORY_SEPARATOR` is not `/` and `TERM` isn't `cygwin` or `posix_isatty()` exists and returns false for `STANDARDOUT`. There are some other cases, but these are the interesting ones for this issue.

That leads arcanist to assume it can use ANSI color sequences, but your php doesn't know what to do with them. 

You could test `TERM=xterm arc diff` and if that works set `TERM=xterm` from `~/.bashrc` if that works.

**Comment by dscho** [comment]
@briannipper have you tried @rimrul's suggestion? What was the outcome?

**Comment by DHowett-MSFT** [comment]
>You're probably using the recommended profile for Git Bash from the microsoft/terminal repo. 

(as the Terminal maintainer, I'd be happy to learn that a bunch of our problems could be resolved by recommending something different.)

**Comment by dscho** [comment]
@DHowett-MSFT I haven't checked: do you call `bin\bash.exe -li` or `usr\bin\bash.exe -li`? The former _should_ work correctly. The latter will miss important environment variable settings.

**Comment by briannipper** [comment]
My apologizes for the unusually long delay in following up on this.

Thank you so much @rimrul you suggestion to use TERM=xterm FIXED all my formatting issues. Also, thank you for the additional explanation of the various layers involved, it's quite helpful to see the pieces as you outline them.

After creating the entry in ~/.bashrc all my arc problems have disappeared. :)

**Comment by DHowett-MSFT** [comment]
Our current recommendation is the former, but it looks like we're missing `--login` and friends. I'll fix that.

Thanks!

**Comment by dscho** [comment]
Yep, without `--login`, the environment won't be set up properly.

**Comment by trajano** [comment]
Perhaps it needs to be recompiled with a newer cygwin library as per https://github.com/microsoft/terminal/issues/4809#issuecomment-596249509

**Comment by dscho** [comment]
> Perhaps it needs to be recompiled with a newer cygwin library as per [microsoft/terminal#4809 (comment)](https://github.com/microsoft/terminal/issues/4809#issuecomment-596249509)

The latest [snapshot](https://wingit.blob.core.windows.net/files/index.html) includes an MSYS2 runtime based on Cygwin v3.1.5, so it is safely newer than the suggested v3.1.2. Even Git for Windows v2.27.0 comes with an MSYS2 runtime based on a newer Cygwin (namely, v3.1.4): https://github.com/git-for-windows/build-extra/blob/main/versions/package-versions-2.27.0.txt#L110

**Comment by trajano** [comment]
@dscho just verified with https://gist.github.com/ketan/2866a236596636311d64 and I see the full color palette now for git bash.  
![image](https://user-images.githubusercontent.com/110627/85308245-fbb45580-b47e-11ea-8722-2ab1de6b0ea4.png)
Though the processes called by git bash is still showing the escape sequences
![image](https://user-images.githubusercontent.com/110627/85308370-23a3b900-b47f-11ea-805a-9b8a1645aacd.png)
so it may need more work from the Cygwin end.  Is there an issue opened up on their end that discusses this?

**Comment by dscho** [comment]
It _might_ just be a question of setting `TERM` appropriately.

**Comment by trajano** [comment]
> 
> 
> It _might_ just be a question of setting `TERM` appropriately.

I think it may be something that needs to happen in OpenJDK unless there's a per process handling.

**Comment by luismbo** [comment]
FWIW, here's a minimal example using git-for-windows 2.29.2.windows.3:

```
$ echo -e "$(tput setaf 1)$(tput bold)RED$(tput sgr0)" > red
$ cat red
RED
$ /c/Windows/System32/more red
←[31m←[1mRED←(B←[m
```

Both `more` and `cat` work as expected if invoked from `cmd.exe`, PowerShell or WSL2. Setting `TERM` to `xterm` (changed from `xterm-256color`; not sure who's setting that) doesn't seem to change the result.

**Comment by dscho** [comment]
@luismbo I just tried to your reproducer, and all the Git parts seem to work all right:

![image](https://user-images.githubusercontent.com/127790/130588674-38454b94-3b62-4dfd-b3e6-98ead35abe9e.png)

But indeed, `more` did not work:

```console
$ C:/Windows/system32/more "$(cygpath -aw /tmp/a1)"
←[31m←[1mRED←(B←[m
```

However, at this stage I really have to ask: since neither Windows Terminal nor `more.com` are developed by the Git for Windows project, why report that problem here? Would it not make a ton more sense to report it with the Windows Terminal project?

**Comment by DHowett** [comment]
This one is interesting! `more`, owing to its age, probably never sets the console mode to include `ENABLE_VIRTUAL_TERMINAL_PROCESSING` (which is not a default flag, but is enabled by CMD and PowerShell).

If MSYS is using ConPTY and running `more` in an isolated console session, it is quite likely that it's inheriting the default/new empty console state. This behavior isn't unexpected, honestly -- I would have been surprised to see `more` opt in to newer console features.

**Comment by luismbo** [comment]
@DHowett, @dscho I used `more` for the sake of easy reproducibility, but I actually came across this issue using my own CLI program. I will investigate how to set the console mode and report back. Thanks for the hint, @DHowett. (FWIW, without any sort of knowledge about the constraints involved, I’m inclined to side with @dscho and suggest that perhaps that mode could be set by default by the Windows Terminal. 🤷‍♂️)

**Comment by DHowett** [comment]
> perhaps that mode could be set by default by the Windows Terminal.

Ah! We beat you by [about two years](https://github.com/microsoft/terminal/pull/2824); however, any application running _inside_ Terminal is allowed to clear that flag if they want, and any sessions spawned by MSYS are sufficiently detached from the original hosting terminal that we can't know WT was involved to begin with :smile:

**Comment by luismbo** [comment]
The plot thickens! I should be reunited with my Windows 11 machine some time next week, so I'll try that too.

**Comment by uniphiny** [comment]
Thanks

**Comment by perlun** [comment]
Apologies for resurrecting an old thread, but... I seem to be running into something slightly similar to this with Git for Windows 2.35.1.windows.2.

I have a console application running which emits ANSI colour codes. I'm running this inside the Windows Terminal.

When I run it through a regular `cmd.exe`, it works perfectly:

![image](https://user-images.githubusercontent.com/630613/155013654-b32ff92c-73bb-485b-a6a4-ef4495ec2e06.png)

_However_, when executed via `bash.exe` provided by Git for Windows, all the ANSI sequences are suddenly ignored:

![image](https://user-images.githubusercontent.com/630613/155013753-fd4503bf-b9a0-421f-b6d7-e187b8845262.png)

The Git Bash profile for Windows Terminal was added using the Git for Windows installer. The exact `bash` path being used here is `C:/Program Files/Git/bin/bash.exe -i -l`.

As can be seen above in the shell prompt etc, other ANSI colouring sequences are working perfectly. I also tested with the gist provided at https://gist.github.com/ketan/2866a236596636311d64, and all of this displays flawlessly as far as I can tell.

![image](https://user-images.githubusercontent.com/630613/155013934-24ef89e8-a629-48e0-a8cd-949da429a9fd.png)

I really don't know who to blame 😁 or where to start debugging this. Is this a problem with:

* Git Bash
* Windows Terminal
* my own application (built from https://github.com/perlang-org/perlang/tree/f84b8da6953146148055e035a1760651bb3cb00f)
* the .NET Framework (which it runs on)
* cosmic rays? 🙈 

Any suggestions are greatly appreciated. 🙏

**Comment by knutwannheden** [comment]
Did you try with `/usr/bin/bash.exe` instead of `/bin/bash.exe` as the executable launched by Terminal? Has fixed issues for me in the past, but even so, I still sometimes have issues with ANSI codes under Git Bash in Terminal ☹️

**Comment by dscho** [comment]
@perlun @knutwannheden Have any of you gentle people tried the suggestion to set the environment variable `TERM=xterm`?

**Comment by perlun** [comment]
@dscho Good point, worth trying. 👍 Here are the results.

## `TERM=xterm-256color`

(This seems to be the default in the `bash` provided with Git for Windows)

![image](https://user-images.githubusercontent.com/630613/155386068-edabf738-20a8-4675-b232-60f1eef01a56.png)

## `TERM=xterm`

![image](https://user-images.githubusercontent.com/630613/155386253-49eb758d-a347-4302-bdff-ffbe2c10bf8e.png)

----

> Did you try with `/usr/bin/bash.exe` instead of `/bin/bash.exe` as the executable launched by Terminal?

@knutwannheden Interesting point; there is in fact another `bash.exe` there. 🤔 So if I replace `C:/Program Files/Git/bin/bash.exe -i -l` with `C:/Program Files/Git/usr/bin/bash.exe -i -l`... that changes the prompt above from `MINGW64` to `MSYS`, but I can't seem to get ANSI colours propagated through from my program regardless.

I tested with both `TERM=xterm-256color` and `TERM=xterm` in this case also.

----

@DHowett - you seem to be a (Microsoft?) hacker in the Windows Terminal subsystem. Could this be related to `ENABLE_VIRTUAL_TERMINAL_PROCESSING` somehow not being propagated to the child process in this case?

In case anyone wants a MCRE, the latest build available from https://perlang.org/download/ has the ANSI colours enabled:

1. Run the installer in a Git for Windows-provided `bash` under Windows Terminal
2. Run `~/.perlang/nightly/bin/perlang` afterwards. It works fine in `CMD.EXE`, but fails miserably in Bash for some utterly odd reason.

**Comment by dscho** [comment]
I have another suggestion: can you run the `perlang` command via `winpty`?

**Comment by luismbo** [comment]
FWIW, I have some more input using my minimal example using `more.com` (https://github.com/git-for-windows/git/issues/2483#issuecomment-841823607):

1. Using Windows 11 (version 10.0.22000 Build 22000) and Windows Terminal 1.11.3471.0 or 1.13.10395.0 doesn't help.
2. `winpty more.com red` doesn't help either.
3. powershell within Git's bash also fails. See:

```
PS Z:\temp> more red
RED

PS Z:\temp> & 'C:\Program Files\Git\bin\bash.exe'

luis@thinkpad MINGW64 /z/temp
$ more.com red
←[31m←[1mRED←(B←[m

luis@thinkpad MINGW64 /z/temp
$ powershell

PS Z:\temp> more red
←[31m←[1mRED←(B←[m
 ```

**Comment by luismbo** [comment]
Just in case using `more` makes you suspicious, I can replicate the same behavior using this ad-hoc `cat` compiled with MSYS2's MINGW64 profile:

```c
#include <stdio.h>

int main(int argc, char **argv)
{
        if (argc != 2) { return -1; }
        FILE *f = fopen(argv[1], "r");
        if (f) {
                char buf[512];
                while (fgets(buf, sizeof buf, f))
                        puts(buf);
                fclose(f);
        }
        return 0;
}
```

**Comment by luismbo** [comment]
Setting the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` mode, as hinted by @DHowett in https://github.com/git-for-windows/git/issues/2483#issuecomment-904626793 works:

```c
// red.c
#include <stdio.h>
#include <windows.h>

int main(void)
{
    HANDLE con = GetStdHandle(STD_OUTPUT_HANDLE);
    DWORD mode;
    GetConsoleMode(con, &mode);
    SetConsoleMode(con, mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
    puts("\e[31m\e[1mRED\e(B\e[m\n");
    return 0;
}
```

This `red.exe` outputs a red-colored "RED" everywhere: Command Prompt, Windows Terminal (both powershell and Git's bash), mintty.

This seems to suggest that Git's bash turns `ENABLE_VIRTUAL_TERMINAL_PROCESSING` off at some point, maybe?

**Comment by dscho** [comment]
> This seems to suggest that Git's bash turns `ENABLE_VIRTUAL_TERMINAL_PROCESSING` off at some point, maybe?

It's not really the Bash, it is the MSYS2 runtime, which is a close fork of the Cygwin runtime.

Could you download the build artifact of https://github.com/git-for-windows/msys2-runtime/actions/runs/1961614385 (which essentially gives a preview of the changes that will come with Cygwin v3.3.5), extract the `msys-2.0.dll` file and replace the one in your installed Git's `usr\bin`, then try again?

**Comment by luismbo** [comment]
@dscho same behaviour. `red.exe` works, ad-hoc cat or more.com, don't work.

```
luis@thinkpad MINGW64 /z/temp
$ cat red
RED

luis@thinkpad MINGW64 /z/temp
$ more.com red
←[31m←[1mRED←(B←[m

luis@thinkpad MINGW64 /z/temp
$ ./my-cat.exe red
←[31m←[1mRED←(B←[m

luis@thinkpad MINGW64 /z/temp
$ winpty ./my-cat.exe red
←[31m←[1mRED←(B←[m

luis@thinkpad MINGW64 /z/temp
$ ./red.exe
RED

luis@thinkpad MINGW64 /z/temp
$ md5sum /usr/bin/msys-2.0.dll
97905bfd688e82d4a43d428229c54457 */usr/bin/msys-2.0.dll
```

**Comment by dscho** [comment]
I won't pretend to have time to dig into this deeper. There is a good chance, though, that this is actually an issue in Cygwin itself. If it is, you won't have to rely on me getting back to this issue, you can [report it to the Cygwin project](https://cygwin.com/problems.html) and have much better chances of seeing this resolved soon.

**Comment by luismbo** [comment]
> Is there a workaround for this?

Can you be more specific?

**Comment by perlun** [comment]
FWIW, my problem described in https://github.com/git-for-windows/git/issues/2483#issuecomment-1049104749 is no loner present. The relevant versions:

- Windows Terminal 1.16.10261.0
- Git Bash 5.2.15(1)-release (from Git for Windows 2.40.0 64-bit)
- Windows 10

I am unsure if this is because of a fix in either one of the above or changes in my own program, but I haven't done any changes that would have an obvious effect on this. (Such changes are being worked on, but they are not yet merged... 😉)

**Comment by Godsmith** [comment]
For anyone having this problem like I just did, I had this issue in Git for Windows 2.34, and it was resolved when I updated to Git for Windows 2.45.1, so it was definitely the Git update that did it.

**Comment by luismbo** [comment]
Probably fixed by this commit: https://github.com/cygwin/cygwin/commit/fdbd1539329ad669606767ab5a63a16f825b4c45
